string.c

```diff
@@ -4948,21 +4948,92 @@ rb_str_insert(VALUE str, VALUE idx, VALUE str2)
 
 /*
  *  call-seq:
- *     str.slice!(integer)           -> new_str or nil
- *     str.slice!(integer, integer)   -> new_str or nil
- *     str.slice!(range)            -> new_str or nil
- *     str.slice!(regexp)           -> new_str or nil
- *     str.slice!(other_str)        -> new_str or nil
- *
- *  Deletes the specified portion from <i>str</i>, and returns the portion
- *  deleted.
- *
- *     string = "this is a string"
- *     string.slice!(2)        #=> "i"
- *     string.slice!(3..6)     #=> " is "
- *     string.slice!(/s.*t/)   #=> "sa st"
- *     string.slice!("r")      #=> "r"
- *     string                  #=> "thing"
+ *    string.slice!(index) -> new_string or nil
+ *    string.slice!(start, length) -> new_string or nil
+ *    string.slice!(range) -> new_string or nil
+ *    string.slice!(regexp, capture = 0) -> new_string or nil
+ *    string.slice!(other_string) -> new_string or nil
+ *
+ *  Deletes and returns the substring of +self+ specified by the arguments.
```

Comment:
I would say that this substring that `slice!` deletes is the same substring that would be matched by `slice`/`[]`, and returns `nil` in the same cases where `slice`/`[]` would return `nil`.  This basically just repeats the `slice`/`[]` documentation and modifies it to show the mutation.  It's basically the same situation as #3609.

If there are cases where `slice!` operates differently than `slice` other than mutating the receiver, I'm guessing those would be considered bugs, but if not, we could document those cases.

Reply:
Changes to #slice! backed out.